### PR TITLE
fix: DetectorVolume containment check

### PR DESCRIPTION
**Pull request by ssdetlab**
The PR makes changes to the DetectorVolume code that address the issue of the surfaces/subvolumes containment checks, i.e. making sure that they are fully inside the volume. Additionally, minor fixes are made in the Json-related parts of the code (see below).

DetectorVolume checkContainment call is moved from the constructor to the end of the "construct" method, where the volume's portals are initialized. Because the containment checks rely on the volume having the initialized Extent and the Extent is derived from the volume's portals, keeping the call in the constructor prevents the proper check from happening, as there are yet no portals to derive the Extent from. Additionally, assert is replaced with invalid_argument thrown in case the internals are not contained. 

To check the compatibility of the external and internal extents a new "canonicalBinning" method is introduced to the VolumeBounds. It returns the binning values that should be enough to describe the relevant parts of the Extent for a given volume shape, e.g.  X,Y,Z bins are returned for cubic volumes, as, for example, there is no reason to check the binMag. Some of the VolumeBounds return the bounding box binning for now as a placeholder. 

A UnitTest is added to the DetectorVolumeTests that checks incompatibility with internals that are not contained. 

A minor fixes had to be introduced to DetectorJsonConverter with respect to the geometry parameters. 

In Transform3 Json writer the default option "transpose" is changed to be false to be consistent with the default reading options.

### CODE REVIEW

**Comment by ssdetlab** [comment]
Forgot to add. The number of segments for polyhedronRepresentation is hardcoded to be 1k. The default option of 1 didn't work with cylindrical shapes on more complex side -- the Extent that was returned for what I think was a barrel/endcap had big enough errors in phi/radius for the check to fail

**Comment by ssdetlab** [comment]
Moved the check inside the DetectorVolumeFactory::construct(). The one that takes the surfaces and subvolumes. Also, made the number of polyhedron segments a parameter with a default value

File: Core/include/Acts/Detector/DetectorVolume.hpp
Comment (asalzburger) [inline Core/include/Acts/Detector/DetectorVolume.hpp:420]: This will make the test extremely slow, I understand the intention, of course ...

Comment (asalzburger) [reply Core/include/Acts/Detector/DetectorVolume.hpp:420]: Here's an idea:
- we could make the containment check optional, e.g. if `nseg = -1` then now checking is done, so in a production workflow (after you have checked your geometry) you can skip the test in each job ?

What do you think?

File: Core/include/Acts/Geometry/GenericCuboidVolumeBounds.hpp
Comment (asalzburger) [inline Core/include/Acts/Geometry/GenericCuboidVolumeBounds.hpp:96]: Do you want the `(x,y,z)` into the base class, so it is clear that classes that do now overload this get just the basic checks?

File: Core/include/Acts/Geometry/TrapezoidVolumeBounds.hpp
Comment (asalzburger) [inline Core/include/Acts/Geometry/TrapezoidVolumeBounds.hpp:147]: Same comment applies as above.

File: Core/include/Acts/Geometry/VolumeBounds.hpp
Comment (asalzburger) [inline Core/include/Acts/Geometry/VolumeBounds.hpp:122]: Here we could give a default (x,y,z) and document that this is what you get per default, and that it can be overloaded.

File: Plugins/Json/include/Acts/Plugins/Json/AlgebraJsonConverter.hpp
Comment (asalzburger) [inline Plugins/Json/include/Acts/Plugins/Json/AlgebraJsonConverter.hpp:31]: Ha, interesting, there is a difference between `detray` style writing - and core writing.

**Review by asalzburger — CHANGES_REQUESTED** [review]
A few comments.

**Comment by ssdetlab** [comment]
Made minor adjustments to the geometry inside the DD4hep Cylindrical Detector test

File: Core/include/Acts/Geometry/VolumeBounds.hpp
Comment (ssdetlab) [reply Core/include/Acts/Geometry/VolumeBounds.hpp:122]: Yes, it'll make the code a bit cleaner. Will implement now

**Review by asalzburger — DISMISSED** [review]

**Review by asalzburger — APPROVED** [review]
Re-approving.

File: Core/include/Acts/Geometry/VolumeBounds.hpp
Comment (ssdetlab) [reply Core/include/Acts/Geometry/VolumeBounds.hpp:126]: Not sure if I understand. I don't think it can be, it does not fall under the special member function (e.g. it's not a constructor).

File: Core/include/Acts/Geometry/CutoutCylinderVolumeBounds.hpp
Comment (ssdetlab) [reply Core/include/Acts/Geometry/CutoutCylinderVolumeBounds.hpp:121]: Fair, will commit changes

Comment (ssdetlab) [reply Core/include/Acts/Geometry/CutoutCylinderVolumeBounds.hpp]: On the other hand, the vector is not a private member. It seems to me that making it one will make the overrides messy, so let's leave it like that for now

**Comment by ssdetlab** [comment]
@asalzburger, looks like it is still waiting for your approval

**Comment by ssdetlab** [comment]
@asalzburger, Just a reminder that the policy-bot is still waiting for the approval

**Review by asalzburger — APPROVED** [review]

**Comment by andiwand** [comment]
we need a codecov overwrite here @asalzburger @paulgessinger

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [d244dee611dc50d6a84564b536228846c0d235a6]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007421)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007420)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007419)||
|:green_circle: | [test_ActsConversionWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007418)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007417)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007416)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007415)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007414)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007413)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007412)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007411)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007410)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007409)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007408)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007407)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007406)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007405)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007404)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007403)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007402)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007401)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007400)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007399)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007398)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37007397)||

**Pull request merged**
